Ensure existing spaces in attribute selectors are valid

RobinMalfait opened the pull request:
This PR fixes an issue where spaces in a selector generated invalid CSS.

Lightning CSS will throw those incorrect lines of CSS out, but if you are in an environment where Lightning CSS doesn't run then invalid CSS is generated.

Given this input:

```html
data-[foo_=_"true"]:flex
```

This will be generated:

```css
.data-\[foo_\=_\"true\"\]\:flex[data-foo=""true] {
  display: flex;
}
```

With this PR in place, the generated CSS will now be:

```css
.data-\[foo_\=_\"true\"\]\:flex[data-foo="true"] {
  display: flex;
}
```

RobinMalfait committed:
ensure existing spaces in attribute selector are valid

If you inject existing spaces in the attribute selector, then invalid
CSS was generated. Lightning CSS itself throws it out, but if you are in
an environment where Lightning CSS doesn't run then invalid CSS exists
in your final CSS file.

This fixes that by ensuring that the spaces around the `=` don't cause
any issues.

RobinMalfait commented:
* **#14703** <a href="https://app.graphite.dev/github/pr/tailwindlabs/tailwindcss/14703?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `next`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @RobinMalfait and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

RobinMalfait commented on packages/tailwindcss/src/variants.ts:
```diff
@@ -901,8 +901,10 @@ export function createVariants(theme: Theme): Variants {
   if (value.includes('=')) {
     value = value.replace(/(=.*)/g, (_fullMatch, match) => {
+      let value = match.slice(1).trim()
+
       // If the value is already quoted, skip.
-      if (match[1] === "'" || match[1] === '"') {
+      if (value[0] === "'" || value[0] === '"') {
```
`_` values in the value are already converted to ` ` which means that we can just trim the valid (if we ignore the first `=` character)

adamwathan commented:
Do we actually want people to be able to write stuff like `data-[foo_=_"true"]:flex` in the first place? You can't put spaces there in CSS so I don't know that it's something we need to support in Tailwind either. I think I'd rather just toss the candidate personally but maybe that's more work?

RobinMalfait commented:
@adamwathan I noticed it while working on the codemods and created an internal issue so that I didn't forget about it. Fun thing is, that you can put spaces in that position in CSS (see https://jsfiddle.net/ezu8qjdk/) and since arbitrary values should be valid CSS (as an escape hatch if the value isn't used), I figured that it is best to at least generate the correct CSS if you put it in that way.

We can of course throw it out (or even fix it at a codemod level), but since it's valid CSS it might be good to keep it? I'm happy with either solution, just want to fix the scenario that exists today where the currently generated CSS is invalid.

RobinMalfait commented:
The part I'm missing is that you can't put spaces if you use `[data-foo ^ = "bar"]` (but `[data-foo ^= "bar"]` is valid). Will add an explicit test for this case.

RobinMalfait committed:
update changelog

philipp-spiess commented on packages/tailwindcss/src/variants.test.ts:
```diff
@@ -2155,6 +2162,14 @@ test('data', async () => {
+      display: flex;
+    }
+
+    .data-\\[potato_\\^\\=_\\"salad\\"\\]\\:flex[data-potato^="salad"] {
```
Is it lightningcss that did remove the spaces again here or did we? I feel like if the argument is "we want to support arbitrary CSS because CSS would also allow that", we should also make sure we output the spaces again in the resulting CSS. Conceptually we just pass-through even though we do auto-quote where we have to.

RobinMalfait replied:
Yep Lightning CSS is cleaning that up for us: https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A6225920%7D%2C%22include%22%3A0%2C%22exclude%22%3A0%2C%22source%22%3A%22.data-%5C%5C%5Bpotato_%5C%5C%5E%5C%5C%3D_%5C%5C%5C%22salad%5C%5C%5C%22%5C%5C%5D%5C%5C%3Aflex%5Bdata-potato%20%5E%3D%20%5C%22salad%5C%22%5D%20%7B%5Cn%20%20color%3A%20red%3B%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D


I don't think re-emitting the _exact_ space is necessary (even though we do). This is what we generate:
```css
.data-\\[potato_\\^\\=_\\"salad\\"\\]\\:flex {
  &[data-potato ^= "salad"] {
    display: flex;
  }
}
```

Only thing I care about is that if you pass in valid CSS that valid CSS comes out and that we didn't generate invalid CSS along the way.

adamwathan commented:
> Fun thing is, that you can put spaces in that position in CSS

Well then…

![continue-tv-land-gif-by-impastor](https://github.com/user-attachments/assets/4e72eb88-e865-40e9-a246-7d21cd155a33)

adamwathan commented on packages/tailwindcss/src/variants.ts:
```diff
@@ -900,9 +906,16 @@ export function createVariants(theme: Theme): Variants {
 function quoteAttributeValue(value: string) {
   if (value.includes('=')) {
+    // Do not allow invalid operators in attribute values
+    if (/[*$^|~]\s+=/.test(value)) {
```
Should we try to always precompile regular expressions?

RobinMalfait replied:
That's something we could do, it's also a non-global one (`/g`) so it's stateless.

This only ever runs in `data-*` or `aria-*` variants when using arbitrary values that contain an `=` so I'm assuming this will be used very rarely. Especially because the `aria-foobar` already translates to `aria-[foobar="true"]`, but this doesn't go via this function since it's a bare value instead of an arbitrary value.

adamwathan replied:
I think this condition would break this variant:

```
data-[potato="^_="]
```

Feeling like the better overall solution to this problem should rely on `segment` — do we just need to segment the value on `=` and quote the last chunk if it isn't already quoted? Maybe we don't need to protect people from writing `data-[potato_^_=_foo]` and just let that be user error that generates junk.

adamwathan commented on packages/tailwindcss/src/variants.ts:
```diff
@@ -900,9 +906,16 @@ export function createVariants(theme: Theme): Variants {
 function quoteAttributeValue(value: string) {
   if (value.includes('=')) {
+    // Do not allow invalid operators in attribute values
+    if (/[*$^|~]\s+=/.test(value)) {
```
I think this condition would break this variant:

```
data-[potato="^_="]
```

Feeling like the better overall solution to this problem should rely on `segment` — do we just need to segment the value on `=` and quote the last chunk if it isn't already quoted? Maybe we don't need to protect people from writing `data-[potato_^_=_foo]` and just let that be user error that generates junk.

RobinMalfait replied:
Reworked it by using `segment` instead of the regex.

RobinMalfait committed:
rely on `segment` when quoting attributes

RobinMalfait committed:
do not rely on regex

This test will verify that the usage of `segment` is correct. Before we
were relying on a regex and check for `(=.*)` which would also match
_inside_ of the quoted value.

philipp-spiess commented on CHANGELOG.md:
```diff
@@ -11,6 +11,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Fixed
+
+- Ensure existing spaces in attribute selectors are valid ([#14703](https://github.com/tailwindlabs/tailwindcss/pull/14703))
```
```suggestion
- Support spaces around attribute selectors in arbitrary variants ([#14703](https://github.com/tailwindlabs/tailwindcss/pull/14703))
```

Maybe this is a bit more clear? WDYT?

adamwathan replied:
Agree with @philipp-spiess, maybe something like this?

```suggestion
- Allow spaces spaces around operators in attribute selector variants ([#14703](https://github.com/tailwindlabs/tailwindcss/pull/14703))
```

"Arbitrary variants" is specifically stuff like `[&>.foo]` whereas `data-[dessert_=_potato]` is an arbitrary value for a functional variant.

philipp-spiess reviewed (approved):
Looks good now and feels like a simple fix. Just one idea regarding the change log entry.

adamwathan commented on packages/tailwindcss/src/variants.ts:
```diff
@@ -898,32 +898,35 @@ export function createVariants(theme: Theme): Variants {
-        ) {
-          return `="${match.slice(1, -2)}" ${match[match.length - 1]}`
-        }
+    // Handle regex flags on unescaped values
```
I know this isn't new to this PR but this comment confused me:

```suggestion
    // Handle case sensitivity flags on unescaped values
```

adamwathan commented on packages/tailwindcss/src/variants.ts:
```diff
@@ -898,32 +898,35 @@ export function createVariants(theme: Theme): Variants {
-          return `="${match.slice(1, -2)}" ${match[match.length - 1]}`
-        }
+    // Handle regex flags on unescaped values
+    if (value.length > 1) {
```
Is it important that we are operating on the non-trimmed value here? It's not obvious to me why we need to have both.

RobinMalfait replied:
I ported it as-is, but can indeed simplify it a bit more. Pushing an update 👍

RobinMalfait committed:
Update packages/tailwindcss/src/variants.ts

Co-authored-by: Adam Wathan <adam.wathan@gmail.com>

RobinMalfait committed:
Update CHANGELOG.md

Co-authored-by: Adam Wathan <adam.wathan@gmail.com>

RobinMalfait committed:
always trim the value
